ENIGMAsystem/SHELL/Universal_System/Extensions/Steamworks/game_client/c_leaderboards.cpp

```diff
@@ -0,0 +1,89 @@
+
+  if (c_leaderboards::current_leaderboard_ != NULL) {
+    if (SteamUserStats()->GetLeaderboardName(c_leaderboards::current_leaderboard_) == leaderboard_name) return;
+  }
+
+  c_leaderboards::current_leaderboard_ = NULL;
+
+  SteamAPICall_t steam_api_call = SteamUserStats()->FindOrCreateLeaderboard(
+      leaderboard_name.c_str(), leaderboard_sort_method, leaderboard_display_type);
+
+  if (steam_api_call != 0) {
+    c_leaderboards::m_callResultFindLeaderboard.Set(steam_api_call, this, &c_leaderboards::on_find_leaderboard);
+    c_leaderboards::loading_ = true;
+  }
+}
+
+void c_leaderboards::on_find_leaderboard(LeaderboardFindResult_t* pFindLeaderboardResult, bool bIOFailure) {
+  c_leaderboards::loading_ = false;
+
+  if (!pFindLeaderboardResult->m_bLeaderboardFound || bIOFailure) return;
+
+  c_leaderboards::current_leaderboard_ = pFindLeaderboardResult->m_hSteamLeaderboard;
+
+  DEBUG_MESSAGE("Calling FindOrCreateLeaderboard succeeded.", M_INFO);
+}
+
+bool c_leaderboards::upload_score(const int score,
+                                  const ELeaderboardUploadScoreMethod leaderboard_upload_score_method) {
+  // while (c_leaderboards::loading_)
+  //   ;  // Wait for the callback of FindOrCreateLeaderboard to be invoked
+
+  if (NULL == c_leaderboards::current_leaderboard_) return false;
+
+  SteamAPICall_t steam_api_call = SteamUserStats()->UploadLeaderboardScore(
+      c_leaderboards::current_leaderboard_, leaderboard_upload_score_method, score, NULL, 0);
+  c_leaderboards::m_SteamCallResultUploadScore.Set(steam_api_call, this, &c_leaderboards::on_upload_score);
+
+  return true;
+}
+
+void c_leaderboards::on_upload_score(LeaderboardScoreUploaded_t* pScoreUploadedResult, bool bIOFailure) {
+  if (!pScoreUploadedResult->m_bSuccess || bIOFailure) {
+    DEBUG_MESSAGE("Failed to upload score to leaderboard.", M_ERROR);
+  }
+}
+
+bool c_leaderboards::download_scores(const ELeaderboardDataRequest leaderboard_data_request, const int range_start,
+                                     const int range_end) {
+  // while (c_leaderboards::loading_)
+  //   ;  // Wait for the callback of FindOrCreateLeaderboard to be invoked
+
+  if (NULL == c_leaderboards::current_leaderboard_) return false;
+
+  SteamAPICall_t steam_api_call = SteamUserStats()->DownloadLeaderboardEntries(
+      c_leaderboards::current_leaderboard_, leaderboard_data_request, range_start, range_end);
+  c_leaderboards::m_SteamCallResultDownloadScores.Set(steam_api_call, this, &c_leaderboards::on_download_scores);
+
+  return true;
+}
+
+void c_leaderboards::on_download_scores(LeaderboardScoresDownloaded_t* pLeaderboardScoresDownloaded, bool bIOFailure) {
+  if (bIOFailure) {
+    DEBUG_MESSAGE("Failed to download scores from leaderboard.", M_ERROR);
+    return;
+  }
+
+  DEBUG_MESSAGE("Downloaded scores from leaderboard.", M_INFO);
+
+  // enigma::signal(&enigma::asyncsteamworks_mutex);
+
+  enigma::posted_async_events.push(std::map<std::string, variant>{
```

Comment:
@RobertBColton 

here this callback function will be called when the scores are downloaded from steam servers then we push the DS data into the queue. Note that it's not the best way for pushing it's not the point. That's the behavior according to my understanding:

 - first we call ``steam_upload_score()`` user script
 - the ``steam_upload_score()`` will then call ``find_leaderboard()`` then call ``upload_score()``
 - ``find_leaderboard()`` will initialize a callresult to be called when it's done
 - when the callresult is ready, it will push the data to the queue

Reply:
This was a good start, and later you added a cookie/tagging mechanism to identify the call results.